### PR TITLE
Use timing channel from PointCloud2, if available. 

### PR DESCRIPTION
**Pull request by ojura**

### CODE REVIEW

**Comment by ojura** [comment]
This is #785, reopened.

Example datasets:

https://ferhr-my.sharepoint.com/:f:/g/personal/jorsulic_fer_hr/EpPrJrsM3pxDpNGn22fDSZYBurSX9j6uyRtkVtnGslx-Og?e=4U44nF

`PointCloud2` messages with the timing channel can be prepared like this: 

https://gist.github.com/ojura/ca9d0e50bc9cd68d33fc2316dfb4e1f8
Here we convert Cartographer's `TimedPointCloud` to a `PointCloud2` message with the timing channel.

@gaschler
> Did you record the data yourself or is it converted from a public dataset?

I converted the data from a private dataset. The owner of the dataset agreed to make it publicly available, provided they remain anonymous.

> IMU timestamps seem very accurate in this bag file ;-)

I am not acquainted with the hardware specifics (e.g. the IMU is probably triggered from a common clock source), but it's an expensive fiber optic gyro. It better be accurate :)

File: cartographer_ros/cartographer_ros/msg_conversion.cc [hunk 1]
```diff
@@ -183,29 +219,82 @@ ToPointCloudWithIntensities(const sensor_msgs::MultiEchoLaserScan& msg) {
...
     }
   }
-  return std::make_tuple(point_cloud, FromRos(message.header.stamp));
+  ::cartographer::common::Time timestamp = FromRos(msg.header.stamp);
+  if (!point_cloud.empty()) {
+    const double duration = point_cloud.back()[3];
+    timestamp += cartographer::common::FromSeconds(duration);
+    for (Eigen::Vector4f& point : point_cloud) {
+      point[3] -= duration;
+      CHECK_LE(point[3], 0) << "Encountered a point with a larger stamp than the last point in the cloud.";
```
Comment (ojura) [inline cartographer_ros/cartographer_ros/msg_conversion.cc]: Maybe you'll want to soften this `CHECK`. Let me know.

File: cartographer_ros/cartographer_ros/sensor_bridge.cc [hunk 2]
```diff
@@ -222,6 +218,9 @@ void SensorBridge::HandleLaserScan(
 void SensorBridge::HandleRangefinder(
     const std::string& sensor_id, const carto::common::Time time,
     const std::string& frame_id, const carto::sensor::TimedPointCloud& ranges) {
+  if (!ranges.empty()) {
+    CHECK_LE(ranges.back()[3], 0);
```
Comment (ojura) [inline cartographer_ros/cartographer_ros/sensor_bridge.cc]: This one as well.

**Comment by gaschler** [comment]
@ojura 
Thanks for uploading example data.
I tested both bag files successfully.

> I converted the data from a private dataset. The owner of the dataset agreed to make it publicly available, provided they remain anonymous.

Is there a chance you have a bag file you created yourself (even a small one) and you would like to open source under a permissive license?

File: cartographer_ros/cartographer_ros/sensor_bridge.cc (hunk 2 above)
Comment (gaschler) [reply cartographer_ros/cartographer_ros/sensor_bridge.cc]: I think it's good to check early.

File: cartographer_ros/cartographer_ros/msg_conversion.cc [hunk 3]
```diff
@@ -32,13 +32,49 @@
...
 #include "ros/ros.h"
 #include "ros/serialization.h"
 #include "sensor_msgs/Imu.h"
 #include "sensor_msgs/LaserScan.h"
 #include "sensor_msgs/MultiEchoLaserScan.h"
 #include "sensor_msgs/PointCloud2.h"
 
+namespace {
+
+struct PointXYZT {
```
Comment (gaschler) [inline cartographer_ros/cartographer_ros/msg_conversion.cc]: add a comment that these must contain 4n floats for alignment

File: cartographer_ros/cartographer_ros/msg_conversion.cc [hunk 4]
```diff
@@ -183,29 +219,85 @@ ToPointCloudWithIntensities(const sensor_msgs::MultiEchoLaserScan& msg) {
...
   // We check for intensity field here to avoid run-time warnings if we pass in
   // a PointCloud2 without intensity.
-  if (PointCloud2HasField(message, "intensity")) {
-    pcl::PointCloud<pcl::PointXYZI> pcl_point_cloud;
-    pcl::fromROSMsg(message, pcl_point_cloud);
+  if (PointCloud2HasField(msg, "intensity")) {
+    if (PointCloud2HasField(msg, "time")) {
+      pcl::PointCloud<PointXYZIT> pcl_point_cloud;
+      pcl::fromROSMsg(msg, pcl_point_cloud);
+      for (const auto& point : pcl_point_cloud) {
```
Comment (gaschler) [inline cartographer_ros/cartographer_ros/msg_conversion.cc]: add point_cloud.points/intensities.reserve() for efficiency

File: cartographer_ros/cartographer_ros/msg_conversion.cc [hunk 5]
```diff
@@ -183,29 +219,85 @@ ToPointCloudWithIntensities(const sensor_msgs::MultiEchoLaserScan& msg) {
...
+      CHECK_LE(point[3], 0) << "Encountered a point with a larger stamp than "
+                               "the last point in the cloud.";
+    }
+  }
+  return std::make_tuple(point_cloud, timestamp);
+}
+
+std::tuple<::cartographer::sensor::TimedPointCloud,
+           ::cartographer::common::Time>
+ToTimedPointCloud(const sensor_msgs::PointCloud2& msg) {
```
Comment (gaschler) [inline cartographer_ros/cartographer_ros/msg_conversion.cc]: Drop this function. The only caller is SensorBridge::HandlePointCloud2Message and can use ToPointCloudWithIntensities instead.

**Comment by ojura** [comment]
> I tested both bag files successfully.

There were three bags :-)

> Is there a chance you have a bag file you created yourself (even a small one) and you would like to open source under a permissive license?

I have some Velodynes laying around here, but I don't think the publicly available Velodyne ROS drivers give you per-point timing information (I am not sure). However, if someone were to extend them to compute the per-point times, these could easily be put into into an additional `PointCloud2` channel, like I am trying to propose in this PR.

But, as I previously mentioned, I have received permission from the person who recorded the datasets to put into public domain these ROS bags that I created from their data (which already had per-point times computed, so I just used that).

**Comment by ojura** [comment]
@gaschler updated

**Comment by ojura** [comment]
Thank you for reviewing, and sorry for being so slow with preparing the datasets :)

**Comment by gaschler** [comment]
@wally-the-cartographer merge

**Comment by gaschler** [comment]
> There were three bags :-)

That one also worked.

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user gaschler. Merge queue now has a length of 1.

File: cartographer_ros/cartographer_ros/msg_conversion.cc [hunk 6]
```diff
@@ -32,13 +32,44 @@
...
+POINT_CLOUD_REGISTER_POINT_STRUCT(
+    PointXYZT, (float, x, x)(float, y, y)(float, z, z)(float, time, time))
+
+POINT_CLOUD_REGISTER_POINT_STRUCT(
+    PointXYZIT,
+    (float, x, x)(float, y, y)(float, z, z)(float, intensity,
+                                            intensity)(float, time, time))
+
 namespace cartographer_ros {
 namespace {
```
Comment (ojura) [inline cartographer_ros/cartographer_ros/msg_conversion.cc]: By the way, I don't think this is true:

https://github.com/googlecartographer/cartographer_ros/blob/82b5eb96884b48f4e77672fdd8d169bac9882d26/cartographer_ros/cartographer_ros/msg_conversion.cc#L76-L79

If you play the bags I provided and visualize the lidar topics, rViz will happily show the point clouds, even though the last float is the [time offset](https://gist.github.com/ojura/ca9d0e50bc9cd68d33fc2316dfb4e1f8#file-timed_pointcloud2-cpp-L41), not [`1.0f`](https://github.com/googlecartographer/cartographer_ros/blob/82b5eb96884b48f4e77672fdd8d169bac9882d26/cartographer_ros/cartographer_ros/msg_conversion.cc#L198).

**Pull request merged**